File: School-F-Backend/src/main/java/com/schoolfurniture/service/ProductService.java
package com.schoolfurniture.service;

import com.schoolfurniture.entity.Product;
// Removed enum imports - now using String fields
import com.schoolfurniture.repository.ProductRepository;
import java.util.ArrayList;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class ProductService {
    
    @Autowired
    private ProductRepository productRepository;
    
    /**
     * Get all active products
     */
    @Transactional(readOnly = true)
    public List<Product> getAllActiveProducts() {
        return productRepository.findByIsActiveTrue();
    }
    
    /**
     * Get all active products with pagination
     */
    @Transactional(readOnly = true)
    public Page<Product> getAllActiveProducts(Pageable pageable) {
        return productRepository.findByIsActiveTrue(pageable);
    }
    
    /**
     * Get all products (including inactive ones) with pagination - for admin use
     */
    @Transactional(readOnly = true)
    public Page<Product> getAllProducts(Pageable pageable) {
        return productRepository.findAll(pageable);
    }
    
    /**
     * Get product by ID
     */
    @Transactional(readOnly = true)
    public Optional<Product> getProductById(Integer productId) {
        return productRepository.findById(productId);
    }
    
    /**
     * Create new product
     */
    public Product createProduct(Product product) {
        validateProduct(product);
        return productRepository.save(product);
    }
    
    /**
     * Update existing product
     */
    public Product updateProduct(Integer productId, Product productDetails) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + productId));
        
        product.setName(productDetails.getName());
        product.setDescription(productDetails.getDescription());
        product.setPrice(productDetails.getPrice());
        product.setStockQuantity(productDetails.getStockQuantity());
        product.setLowStockThreshold(productDetails.getLowStockThreshold());
        product.setImageUrl(productDetails.getImageUrl());
        product.setIsActive(productDetails.getIsActive());
        product.setCategory(productDetails.getCategory());
        product.setColor(productDetails.getColor());
        
        validateProduct(product);
        return productRepository.save(product);
    }
    
    /**
     * Delete product (soft delete by setting isActive to false)
     */
    public void deleteProduct(Integer productId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + productId));
        
        product.setIsActive(false);
        productRepository.save(product);
    }
    
    /**
     * Search products by name or description
     */
    @Transactional(readOnly = true)
    public List<Product> searchProducts(String searchTerm) {
        return productRepository.searchProducts(searchTerm);
    }
    
    /**
     * Search products with pagination
     */
    @Transactional(readOnly = true)
    public Page<Product> searchProducts(String searchTerm, Pageable pageable) {
        return productRepository.searchProducts(searchTerm, pageable);
    }
    
    /**
     * Get products by category
     */
    @Transactional(readOnly = true)
    public List<Product> getProductsByCategory(String category) {
        return productRepository.findByCategory(category);
    }
    
    /**
     * Get products by category with pagination
     */
    @Transactional(readOnly = true)
    public Page<Product> getProductsByCategory(String category, Pageable pageable) {
        return productRepository.findByCategoryAndIsActiveTrue(category, pageable);
    }
    
    /**
     * Get products by color
     */
    @Transactional(readOnly = true)
    public List<Product> getProductsByColor(String color) {
        return productRepository.findByColor(color);
    }
    
    /**
     * Get products by color with pagination
     */
    @Transactional(readOnly = true)
    public Page<Product> getProductsByColor(String color, Pageable pageable) {
        return productRepository.findByColorAndIsActiveTrue(color, pageable);
    }
    
    /**
     * Get products by price range
     */
    @Transactional(readOnly = true)
    public List<Product> getProductsByPriceRange(BigDecimal minPrice, BigDecimal maxPrice) {
        return productRepository.findByPriceRange(minPrice, maxPrice);
    }
    
    /**
     * Get products by price range with pagination
     */
    @Transactional(readOnly = true)
    public Page<Product> getProductsByPriceRange(BigDecimal minPrice, BigDecimal maxPrice, Pageable pageable) {
        return productRepository.findByPriceRange(minPrice, maxPrice, pageable);
    }
    
    /**
     * Advanced search with multiple filters
     */
    @Transactional(readOnly = true)
    public Page<Product> searchProductsWithFilters(String searchTerm, String category, String color, 
                                                   BigDecimal minPrice, BigDecimal maxPrice, Pageable pageable) {
        return productRepository.findProductsWithFilters(searchTerm, category, color, minPrice, maxPrice, pageable);
    }
    
    /**
     * Find products by category name
     */
    public List<Product> findByCategoryName(String categoryName) {
        return productRepository.findByCategory(categoryName);
    }
    
    /**
     * Find products by color name
     */
    public List<Product> findByColorName(String colorName) {
        return productRepository.findByColor(colorName);
    }
    
    /**
     * Find product by code article
     */
    public Optional<Product> findByCodeArticle(String codeArticle) {
        return productRepository.findByCodeArticle(codeArticle);
    }
    
    /**
     * Get low stock products
     */
    @Transactional(readOnly = true)
    public List<Product> getLowStockProducts() {
        return productRepository.findLowStockProducts();
    }
    
    /**
     * Get out of stock products
     */
    @Transactional(readOnly = true)
    public List<Product> getOutOfStockProducts() {
        return productRepository.findOutOfStockProducts();
    }
    
    /**
     * Get most popular products
     */
    @Transactional(readOnly = true)
    public List<Object[]> getMostPopularProducts(Pageable pageable) {
        return productRepository.findMostPopularProducts();
    }
    
    /**
     * Get recently added products
     */
    @Transactional(readOnly = true)
    public List<Product> getRecentlyAddedProducts(Pageable pageable) {
        return productRepository.findRecentlyAddedProducts(pageable);
    }
    
    /**
     * Update product stock
     */
    public Product updateProductStock(Integer productId, Integer newStockQuantity) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + productId));
        
        if (newStockQuantity < 0) {
            throw new RuntimeException("Stock quantity cannot be negative");
        }
        
        product.setStockQuantity(newStockQuantity);
        return productRepository.save(product);
    }
    
    /**
     * Reduce product stock (for order processing)
     */
    public Product reduceProductStock(Integer productId, Integer quantity) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + productId));
        
        if (product.getStockQuantity() < quantity) {
            throw new RuntimeException("Insufficient stock for product: " + product.getName());
        }
        
        product.setStockQuantity(product.getStockQuantity() - quantity);
        return productRepository.save(product);
    }
    
    /**
     * Increase product stock (for order cancellation or return)
     */
    public Product increaseProductStock(Integer productId, Integer quantity) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + productId));
        
        product.setStockQuantity(product.getStockQuantity() + quantity);
        return productRepository.save(product);
    }
    
    /**
     * Get product statistics for admin dashboard
     */
    @Transactional(readOnly = true)
    public Object[] getProductStatistics() {
        return productRepository.getProductStatistics();
    }
    
    /**
     * Validate product data
     */
    private void validateProduct(Product product) {
        if (product.getName() == null || product.getName().trim().isEmpty()) {
            throw new RuntimeException("Product name is required");
        }
        
        if (product.getPrice() == null || product.getPrice().compareTo(BigDecimal.ZERO) <= 0) {
            throw new RuntimeException("Product price must be greater than 0");
        }
        
        if (product.getStockQuantity() == null || product.getStockQuantity() < 0) {
            throw new RuntimeException("Stock quantity cannot be negative");
        }
        
        if (product.getCategory() == null) {
            throw new RuntimeException("Product category is required");
        }
        
        // Category and color validation is handled by enum constraints
    }
    
    /**
     * Get distinct categories from database
     */
    @Transactional(readOnly = true)
    public List<String> getDistinctCategories() {
        return productRepository.findDistinctCategories();
    }
    
    /**
     * Get distinct colors from database
     */
    @Transactional(readOnly = true)
    public List<String> getDistinctColors() {
        return productRepository.findDistinctColors();
    }
}

File: School-F-Front/school-furniture-angular/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { 
  Product, 
  ProductResponse, 
  ProductRequest, 
  PagedResponse, 
  Category, 
  Color,
  CategoryResponse,
  ColorResponse
} from '../models';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private readonly baseUrl = 'http://localhost:8081/api';

  constructor(private http: HttpClient) {}

  // Get all products with pagination
  getAllProducts(page: number = 0, size: number = 12, sortBy: string = 'name', sortDir: string = 'asc'): Observable<PagedResponse<ProductResponse>> {
    const params = new HttpParams()
      .set('page', (page ?? 0).toString())
      .set('size', (size ?? 12).toString())
      .set('sortBy', sortBy || 'name')
      .set('sortDir', sortDir || 'asc');
    
    return this.http.get<PagedResponse<ProductResponse>>(`${this.baseUrl}/products`, { params });
  }

  // Get product by ID
  getProductById(id: number): Observable<ProductResponse> {
    return this.http.get<ProductResponse>(`${this.baseUrl}/products/${id}`);
  }

  // Search products
  searchProducts(query: string, page: number = 0, size: number = 12, sortBy: string = 'name', sortDir: string = 'asc'): Observable<PagedResponse<ProductResponse>> {
    const params = new HttpParams()
      .set('query', query || '')
      .set('page', (page ?? 0).toString())
      .set('size', (size ?? 12).toString())
      .set('sortBy', sortBy || 'name')
      .set('sortDir', sortDir || 'asc');
    
    return this.http.get<PagedResponse<ProductResponse>>(`${this.baseUrl}/products/search`, { params });
  }

  // Filter products by category
  getProductsByCategory(category: Category, page: number = 0, size: number = 12, sortBy: string = 'name', sortDir: string = 'asc'): Observable<PagedResponse<ProductResponse>> {
    const params = new HttpParams()
      .set('category', category || '')
      .set('page', (page ?? 0).toString())
      .set('size', (size ?? 12).toString())
      .set('sortBy', sortBy || 'name')
      .set('sortDir', sortDir || 'asc');
    
    return this.http.get<PagedResponse<ProductResponse>>(`${this.baseUrl}/products/filter/category`, { params });
  }

  // Filter products by color
  getProductsByColor(color: Color, page: number = 0, size: number = 12, sortBy: string = 'name', sortDir: string = 'asc'): Observable<PagedResponse<ProductResponse>> {
    const params = new HttpParams()
      .set('color', color || '')
      .set('page', (page ?? 0).toString())
      .set('size', (size ?? 12).toString())
      .set('sortBy', sortBy || 'name')
      .set('sortDir', sortDir || 'asc');
    
    return this.http.get<PagedResponse<ProductResponse>>(`${this.baseUrl}/products/filter/color`, { params });
  }

  // Filter products by price range
  getProductsByPriceRange(minPrice: number, maxPrice: number, page: number = 0, size: number = 12, sortBy: string = 'name', sortDir: string = 'asc'): Observable<PagedResponse<ProductResponse>> {
    const params = new HttpParams()
      .set('minPrice', (minPrice ?? 0).toString())
      .set('maxPrice', (maxPrice ?? 999999).toString())
      .set('page', (page ?? 0).toString())
      .set('size', (size ?? 12).toString())
      .set('sortBy', sortBy || 'name')
      .set('sortDir', sortDir || 'asc');
    
    return this.http.get<PagedResponse<ProductResponse>>(`${this.baseUrl}/products/filter/price`, { params });
  }

  // Advanced search with multiple filters
  searchProductsWithFilters(
    query?: string,
    category?: string,
    color?: string,
    minPrice?: number,
    maxPrice?: number,
    page: number = 0,
    size: number = 12,
    sortBy: string = 'name',
    sortDir: string = 'asc'
  ): Observable<PagedResponse<ProductResponse>> {
    let params = new HttpParams()
      .set('page', (page ?? 0).toString())
      .set('size', (size ?? 12).toString())
      .set('sortBy', sortBy || 'name')
      .set('sortDir', sortDir || 'asc');

    if (query && query.trim()) {
      params = params.set('query', query.trim());
    }
    if (category && category.trim()) {
      params = params.set('category', category.trim());
    }
    if (color && color.trim()) {
      params = params.set('color', color.trim());
    }
    if (minPrice !== null && minPrice !== undefined) {
      params = params.set('minPrice', (minPrice ?? 0).toString());
    }
    if (maxPrice !== null && maxPrice !== undefined) {
      params = params.set('maxPrice', (maxPrice ?? 999999).toString());
    }

    return this.http.get<PagedResponse<ProductResponse>>(`${this.baseUrl}/products/filter`, { params });
  }

  // Get recent products
  getRecentProducts(limit: number = 10): Observable<ProductResponse[]> {
    const params = new HttpParams().set('limit', (limit ?? 10).toString());
    return this.http.get<ProductResponse[]>(`${this.baseUrl}/products/recent`, { params });
  }

  // Get popular products
  getPopularProducts(limit: number = 10): Observable<ProductResponse[]> {
    const params = new HttpParams().set('limit', (limit ?? 10).toString());
    return this.http.get<ProductResponse[]>(`${this.baseUrl}/products/popular`, { params });
  }

  // Admin methods (require authentication)
  createProduct(product: ProductRequest): Observable<ProductResponse> {
    return this.http.post<ProductResponse>(`${this.baseUrl}/products`, product);
  }

  updateProduct(id: number, product: ProductRequest): Observable<ProductResponse> {
    return this.http.put<ProductResponse>(`${this.baseUrl}/products/${id}`, product);
  }

  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/products/${id}`);
  }

  updateStock(id: number, quantity: number): Observable<ProductResponse> {
    const params = new HttpParams().set('quantity', (quantity ?? 0).toString());
    return this.http.patch<ProductResponse>(`${this.baseUrl}/products/${id}/stock`, null, { params });
  }

  checkAvailability(id: number, quantity: number): Observable<boolean> {
    const params = new HttpParams().set('quantity', (quantity ?? 1).toString());
    return this.http.get<boolean>(`${this.baseUrl}/products/${id}/availability`, { params });
  }

  // Get all categories
  getAllCategories(): Observable<CategoryResponse[]> {
    return this.http.get<CategoryResponse[]>(`${this.baseUrl}/categories`);
  }

  // Get category by name
  getCategoryByName(name: string): Observable<CategoryResponse> {
    return this.http.get<CategoryResponse>(`${this.baseUrl}/categories/${name}`);
  }

  // Get all colors
  getAllColors(): Observable<ColorResponse[]> {
    return this.http.get<ColorResponse[]>(`${this.baseUrl}/colors`);
  }

  // Get color by name
  getColorByName(name: string): Observable<ColorResponse> {
    return this.http.get<ColorResponse>(`${this.baseUrl}/colors/${name}`);
  }
}